h2/src/main/org/h2/jdbcx/JdbcConnectionPool.java

```diff
@@ -239,19 +245,19 @@ private Connection getConnectionNow() throws SQLException {
      *
      * @param pc the pooled connection
      */
-    synchronized void recycleConnection(PooledConnection pc) {
-        if (activeConnections <= 0) {
+    private void recycleConnection(PooledConnection pc) {
+        int active = activeConnections.decrementAndGet();
+        if (active < 0) {
```

Comment:
need to increment before throwing an AssertionError otherwise you leave it in a permanentaly broken state

Reply:
increment added

Reply:
throwing an AssertionError means this situation is impossible or should never happen(I think so), and leaving the PooledConnection not recycled or closed too, maybe the increment of activeConnections is not that important